usr/sbin/p11sak/p11sak.c

```diff
@@ -1910,104 +2051,264 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, p11sak_kt kt, int long_print)
+                    p11_get_ckr(rc));
+            return 1;
+        }
+        if (count == 0)
+            break;
+
+        rc = tok_key_get_key_type(session, hkey, &keyclass, &keytype,
+                &keylength);
+        if (rc != CKR_OK) {
+            printf("Invalid key type (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+            continue;
+        }
+
+        rc = tok_key_get_label_attr(session, hkey, &label);
+        if (rc != CKR_OK) {
+            printf("Retrieval of label failed (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+        }
+
+        if (strcmp(rm_label, "") != 0) {
+            if (strcmp(rm_label, label) == 0) {
+
+                if (*forceAll) {
+                    if (!(buff = malloc(strlen("y")))) {
+                        printf("Error allocating space for buff\n");
+                        return CKR_HOST_MEMORY;
+                    }
+                    buff = strcpy(buff, "y");
+                } else {
+                    printf("Are you sure to delete %s? (y/n)\n", label);
+
+                    nread = getline(&buff, &buflen, stdin);
+                    if (nread == -1) {
+                        rc = -1;
+                        goto done;
+                    }
+                    buff[nread - 1] = '\0';
+                }
+
+                if (strcmp(buff, "y") == 0) {
+                    printf("Destroy Object with Label: %s\n", label);
+                    rc = funcs->C_DestroyObject(session, hkey);
+                    if (rc != CKR_OK) {
+                        printf(
+                                "Key with label %s could not be destroyed (error code 0x%lX: %s)\n",
+                                label, rc, p11_get_ckr(rc));
+
+                        return rc;
+                    }
+                    printf("DONE - Destroy Object with Label: %s\n", label);
+                    break;
+                } else if (strcmp(buff, "n") == 0) {
+                    printf("Skip deleting Key\n");
+                    break;
+                } else {
+                    printf("Please just enter (y) for yes or (n) for no.\n");
+                }
+            }
+        }
+        else {
+            if (*forceAll) {
+                printf("Destroy Object with Label: %s\n", label);
+                rc = funcs->C_DestroyObject(session, hkey);
+                if (rc != CKR_OK) {
+                    printf(
+                            "Key with label %s could not be destroyed (error code 0x%lX: %s)\n",
+                            label, rc, p11_get_ckr(rc));
+                    return rc;
+                }
+                printf("DONE - Destroy Object with Label: %s\n", label);
+            } else {
+                while (1) {
+                    printf("Are you sure to delete %s? (y/n)\n", label);
+
+                    nread = getline(&buff, &buflen, stdin);
```

Comment:
I guess you are intentionally not freeing buff within the loop iteration, so that it is still allocated from the previous iteration here, and is reused/reallocted by getline ?